### PR TITLE
app-modes ground, flight, auto

### PR DESCRIPTION
**Pull request by HeinerTholen**
Add buttons in the bottom of the left drawer to switch the add between three modes: Ground, Flight, Auto. 

"Ground" and "Flight" modes correspond to `satNav.isInFlight` returning false and true, respectively. Mode "Auto" uses `satNav.isInFlight` to determine if the app should be rendered in ground or flight mode.

I thought about making a simple switch in the settings, but this layout offers more convenience in my opinion. Especially balloonists and other slow flying pilots will be able to easily switch modes when starting or landing. Also, I find it nice to keep the app in "Ground" mode, when driving on a train or in a car, where it otherwise switches modes when reaching 30 kt.

Note that in `src/qml/items/MFM.qml` I placed a TODO on line 82. I think that the function can be removed, but I'm not 100% sure.

### CODE REVIEW

File: src/qml/items/MFM.qml [hunk 1]
```diff
@@ -75,10 +75,11 @@ Item {
...
                 if (flightMap.followGPS === true) {
-                    flightMap.bearing = satNav.isInFlight ? satNav.track : 0.0
+                    flightMap.bearing = globalSettings.isAppInFlightMode(satNav.isInFlight) ? satNav.track : 0.0
                 }
             }
 
+            // TODO is this function needed? (onTrackChanged should do it, since the track must have changed if isInFlight changed)
             function onIsInFlightChanged() {
                 if (flightMap.followGPS === true) {
                     flightMap.bearing = satNav.isInFlight ? satNav.track : 0.0
```
Comment (HeinerTholen) [inline src/qml/items/MFM.qml:85]: This is the TODO. Is this function needed?

**Comment by kebekus** [comment]
Dear Heiner,

I had a look at your pull request. Thank you for that! I would like to add the feature in the upcoming release 1.9.0 (so, not in the next release, which I would like to get out as soon as possible). However, I do have a few remarks.

- I feel that the three buttons in the menu drawer are in conflict with the design goal to make the app as simple as possible. I would prefer one single Option "Automatic flight detection" in the Settings menu. I understand that there are cases where you wish to enforce ground mode (e.g. while working with the app in a train or car). However, I do see these are minor use cases that do not justify to make the app design more complicated -- the app is all about flight, after all. If you want to enforce ground mode,  go to the Android "Settings" and disable GPS temporarily (we might want to explain that in the manual).

- There are several places throughout the app where the GUI simplifies in flight mode. For instance, the "Info"-Menu vanishes and the "Library / Map" becomes unavailable to prevent the user from accidentally erasing or updating maps in-flight. Also, the app explicitly asks if you want to "Exit" when in flight. Pilots of slow aircraft will probably switch "flight detection" off permanently (or in the present scheme, enforce flight mode permanently). The GUI simplification should not apply in these cases.

Heiner, I hope you're not mad at me now. Feel free to disagree and discuss.

Best,

Stefan.

**Comment by HeinerTholen** [comment]
Dear Stefan, 

* regarding your first point: I understand that the app should be simple and I absolutely agree. I would argue that the changes are not a significant complication, but as complications go - you're right - the often come in tiny steps and you end up with something complicated. I also understand that the app is about flight. So it's not really neccessary to be able to download maps while on train or car.
* On your second point: If we disable automatic flight detection entirely, then balloonists will neither see the Nav-bar, nor the five-minute bar, ever. That's big miss, in my opinion. As you said above, the app is all about flight. So if you can't enable the flight mode by hand, we can also leave things as they are, because balloons and paramotors hardly reach 30kt at all.

So, in conclusion, in my mind, the big win would be a switch to enable flight mode by hand, not keeping ground-mode at all times.

I have to admit that I do really like those buttons in the menu drawer. But I'm willing to give them up. How about making a settings switch that enables flight mode at all times? What should that be called?

**Comment by acrovato** [comment]
Hi everyone,

I like the idea of letting the user have a full control over what he is doing, but I also understand and agree that the app should be kept as simple as possible.  

Regarding the addition of a mode chooser, I would have proposed a three-way switch allowing to select auto, ground or flight mode. However, I understand that the app is all about flying. Therefore, I agree that adding a simpler switch allowing to cycle between auto and flight modes is sufficient.

Another, simpler, solution would be to lower the flight speed threshold to ~10 kts (taxi speed for aircraft). That would allow slow flyers to reach flight mode without penalizing the other users and while keeping the app as simple as it is now. However, an issue that could arise would be that the app constantly switches between flight and ground modes if the vehicle flies near the threshold speed, or when the satellite connection is bad.

What do you think?

Cheers

**Comment by HeinerTholen** [comment]
Hi Adrien, 

I assume you didn't checkout the branch so here's a screenshot of the "three-way-button" (bottom of the drawer, they are simple tool-buttons). Would you know a better Qt-class for the job? Just out of curiosity...

Bildschirmfoto 2020-07-08 um 15 59 17

I also thought about reducing the speed threshold even lower than 30kts, but I think that it doesn't really solve the problem (at least for ballons) and could potentially be very annoying.

**Comment by HeinerTholen** [comment]
Okay, so instead of the three way switch in the drawer, I'll make a simple settings switch that says "Automatic flight detection":
* "off" has the description text: "App always in flight-mode"
* "on" has the description text: "Switching to flight-mode at 30kts"

@kebekus, if you confirm, I will make it so. =) 

One more thing, though: we are working on the flight-log as well. With the three-way-button in the drawer, it would be quite natural to assume that flight-log entry starts when the user switches the app to flight mode and ends when switching back. This way, ballon and paramotor pilots would benefit from the flight log as well... Just an idea.. maybe a bad one.. let me know what you think.

**Comment by acrovato** [comment]
> I assume you didn't checkout the branch so here's a screenshot of the "three-way-button" (bottom of the drawer, they are simple tool-buttons). Would you know a better Qt-class for the job? Just out of curiosity...

You are assuming right. I am not familiar with Qt, so I do not know which switches or buttons are available, unfortunately. I find your interface clear and intuitive so it's fine for me.

> I also thought about reducing the speed threshold even lower than 30kts, but I think that it doesn't really solve the problem (at least for ballons) and could potentially be very annoying.

I thought balloons would move faster than 10 kts, but yes, reducing the threshold could have unwanted side effects.

**Comment by kebekus** [comment]
Dear all,

this morning I created a mess by accidentally merging this pull request into the master branch (and then pushing). I reverted the commit and pushed again, so the total effect is zero. I apologize for the mess -- I should learn not to do such things before I've had coffee.

Best,

Stefan.

**Comment by HeinerTholen** [comment]
Alright, I just updated the code. The new settings switch is there:

Bildschirmfoto 2020-07-10 um 18 19 06

These are the changes in the app:
* `src/qml/items/MFM.qml`: 
  * For `bearing`, replacing both, `onTrackChanged` and `onIsInFlightChanged` connections to `onLastValidTrackChanged`, since the corresponding signals `trackChanged` and `isInFlightChanged` don't exist anymore. Using `(!globalSettings.autoFlightDetection || satNav.isInFlight)` here.
  * `RotationAnimation`: using `(!globalSettings.autoFlightDetection || satNav.isInFlight)` here.
  * Five-minute-bar: using `(!globalSettings.autoFlightDetection || satNav.isInFlight)` here.
  * Nav-bar: using `(!globalSettings.autoFlightDetection || satNav.isInFlight)` here.
* `src/qml/main.qml`
  * Leaving everything as is. "About" and "Manual" will disappear if going beyond 30kt.
* `src/qml/pages/SettingsPage.qml `:
  * The new switch.
  * Leaving "Maps" as is, ie. when going beyond 30kt, you can't download maps anymore.

**Comment by kebekus** [comment]
Wonderful job! Blazing fast. Thank you so much.

**Comment by HeinerTholen** [comment]
Thanks, my pleasure!
